### PR TITLE
README.md for new changes

### PR DESCRIPTION
**Pull request by supriyopaul**
1. README.md for new changes
2. str2int removed from formatters
3. 'event' key added to LOG_STRUCTURE
4. LOG_STRUCTURE now a class level constant (so that formatters cannot mess with the LOG_STRUCTURE)
5. for more than 2 targets multiprocessing is used
6. MongoDBforwarder can withstand disconnection from `mongod` server (not sure what happens if duration is long...)
**Pending**: MongoDBforwarder's _parse_msg_for_mongoDB should have the capablity of checking if keys containing '.'  which are against mongoDB line protocol

### CODE REVIEW

File: logagg/__init__.py [hunk 1]
```diff
@@ -1,6 +1,8 @@
+import util
```
Comment (supriyopaul) [inline logagg/__init__.py]: Need to think what functions in this we can export to deeputil.

File: logagg/collector.py [hunk 2]
```diff
@@ -61,6 +62,15 @@ def __init__(self, fpaths, nsq_sender,
...
 
+    def validate_log_format(self, log):
+        for key in log:
+            #To avoid duplicate information
+            if key in log and key in log['data']:
+                log[key] = log['data'].pop(key)
+
+            assert bool(key in self.LOG_STRUCTURE)
+            assert isinstance(log[key], self.LOG_STRUCTURE[key])
+
```
Comment (supriyopaul) [inline logagg/collector.py]: - Move redundancy removal to a separate function
- bool() not required
- we need to test whether value is empty or not (are all values required to be non empty?)

File: logagg/collector.py [hunk 3]
```diff
@@ -75,6 +85,7 @@ def collect_log_lines(self, log_file):
                     file=fpath,
                     host=self.HOST,
                     formatter=L['formatter'],
+                    event='default_event',
```
Comment (supriyopaul) [inline logagg/collector.py]: default_event => event

File: logagg/formatters.py [hunk 4]
```diff
@@ -48,30 +50,34 @@ def nginx_access(line):
...
               u'http_referer': u'-',
               u'http_user_agent': u'python-requests/2.18.4',
               u'http_x_forwarded_for': u'-',
               u'remote_addr': u'192.158.0.51',
               u'remote_user': u'-',
               u'request': u'POST /mpub?topic=heartbeat HTTP/1.1',
-              u'request_time': 0.0,
-              u'status': 404.0,
+              u'request_time': u'0.000',
+              u'status': u'404',
```
Comment (supriyopaul) [inline logagg/formatters.py]: type modification must be done only by the formatter function and that too specific to each field without using a general function like str2int or some such.

Comment (supriyopaul) [reply logagg/formatters.py]: for eg: in nginx:

log['request_time'] = float(log['request_time'])

File: logagg/formatters.py [hunk 5]
```diff
@@ -48,30 +50,34 @@ def nginx_access(line):
...
+     'type': 'metric'}
     '''
 #TODO Handle nginx error logs
     log = json.loads(line)
     timestamp_iso = datetime.datetime.utcfromtimestamp(float(log['timestamp'])).isoformat()
     log.update({'timestamp':timestamp_iso})
     if '-' in log.get('upstream_response_time'):
         log['upstream_response_time'] = 0.0
-    log = convert_str2int(log)
+    event = log['request'].split(' ')[0]+'_request'
```
Comment (supriyopaul) [inline logagg/formatters.py]: pep8 failure

File: logagg/formatters.py [hunk 6]
```diff
@@ -150,6 +149,7 @@ def django(line):
                           u'timestamp': 1506061932546,
                           u'user': u'root'},
               'timestamp': '2017-09-22T06:32:15'},
+     'level': 'INFO',
```
Comment (supriyopaul) [inline logagg/formatters.py]: are we testing in validation function in collector that the values of level ought to be one of 'INFO', 'DEBUG' etc?

File: logagg/formatters.py [hunk 7]
```diff
@@ -205,6 +206,7 @@ def basescript(line):
               u'level': u'warning',
               u'timestamp': u'2018-02-07T06:37:00.297610Z',
               u'type': u'log'},
+     'event': u'exited via keyboard interrupt',
```
Comment (supriyopaul) [inline logagg/formatters.py]: the event value should look like a variable (even in examples to set the right expectation to the reader)

File: logagg/formatters.py [hunk 8]
```diff
@@ -230,14 +233,17 @@ def elasticsearch(line):
...
+              'gc_count': '296816',
               'level': 'WARN ',
               'message': 'o.e.m.j.JvmGcMonitorService',
               'plugin': 'Glsuj_2',
               'query_time_ms': 1200.0,
               'resp_time_ms': 1300.0,
               'timestamp': '2017-08-30T06:27:19,158'},
-     'timestamp': '2017-08-30T06:27:19,158'}
+     'event': 'o.e.m.j.JvmGcMonitorService',
+     'level': 'WARN ',
```
Comment (supriyopaul) [inline logagg/formatters.py]: level value has a space at the end. we need to validate value in collector.

File: logagg/formatters.py [hunk 9]
```diff
@@ -264,11 +270,14 @@ def elasticsearch(line):
                 continue
 
         data = dict(zip(keys,values))
-        data = convert_str2int(data)
+        event = data['message']
 
         return dict(
                 timestamp=values[0],
-                data=data
+                level=values[1],
```
Comment (supriyopaul) [inline logagg/formatters.py]: ts, level = values
timestamp = ts
level = level

File: logagg/forwarder.py [hunk 10]
```diff
@@ -38,7 +40,7 @@ def start(self):
 
         # Establish connection to nsq from where we get the logs
         # Since, it is a blocking call we are starting the reader here.
-        self.log.info('Starting nsq reader')
+        self.log.info('Starting_nsq_reader')
```
Comment (supriyopaul) [inline logagg/forwarder.py]: event string should be all lower case. also, should this be a debug log?

File: logagg/forwarder.py [hunk 11]
```diff
@@ -86,12 +88,29 @@ def _ack_messages(self, msgs):
...
-            self.log.debug('Starting to write messeges ', target=one_target)
+    def _send_msgs_to_target(self, target, msgs):
+        while 1:
+            try:
+                target.handle_logs(msgs)
+                break
+            except (SystemExit, KeyboardInterrupt): raise
+            except:
+                # FIXME: need to check whether this works at all
+                # FIXME: do we log the failed messages themselves somewhere?
```
Comment (supriyopaul) [inline logagg/forwarder.py]: need to log issue(s) for these FIXMEs

File: logagg/forwarder.py [hunk 12]
```diff
@@ -86,12 +88,29 @@ def _ack_messages(self, msgs):
...
+                self.log.exception('_send_msgs_to_target_failed',
+                        target=target.__name__, num_msgs=len(msgs))
+                time.sleep(5) # FIXME: don't hard code this number
```
Comment (supriyopaul) [inline logagg/forwarder.py]: need to remove hardcoded number and parametrize at some level (command-line arg or class level const etc)

File: logagg/forwarder.py [hunk 13]
```diff
@@ -86,12 +88,29 @@ def _ack_messages(self, msgs):
...
+                # FIXME: also implement some sort of backoff sleep
 
+    def _write_messages(self, msgs):
+        #import pdb; pdb.set_trace()
```
Comment (supriyopaul) [inline logagg/forwarder.py]: needs to be removed

File: logagg/forwarders.py [hunk 14]
```diff
@@ -25,95 +32,147 @@ class MongoDBForwarder(BaseForwarder):
...
+    @keeprunning(wait_secs=SERVER_SELECTION_TIMEOUT, exit_on_success=True)
     def _ensure_connection(self):
         # Establish connection to MongoDB to store the nsq messages
         url = 'mongodb://%s:%s@%s:%s' % (self.user,
                                          self.passwd,
                                          self.host,
                                          self.port)
         client = MongoClient(url, serverSelectionTimeoutMS=self.SERVER_SELECTION_TIMEOUT)
-        self.log.info('Established connecton to MongoDB server: %s' % (self.host))
+        self.log.info('MongoDB_server_connection_established', host=self.host)
```
Comment (supriyopaul) [inline logagg/forwarders.py]: debug logs here please

File: logagg/forwarders.py [hunk 15]
```diff
@@ -25,95 +32,147 @@ class MongoDBForwarder(BaseForwarder):
...
         self.database = client[self.db_name]
-        self.log.info('Created database: %s at MongoDB' % (self.db_name))
+        self.log.info('MongoDB_database_created', db=self.db_name)
```
Comment (supriyopaul) [inline logagg/forwarders.py]: lower case event names

File: logagg/forwarders.py [hunk 16]
```diff
@@ -25,95 +32,147 @@ class MongoDBForwarder(BaseForwarder):
...
         self.collection = self.database[self.coll]
-        self.log.info('Created collection: %s for MongoDB database %s' % (self.collection, self.db_name))
+        self.log.info('MongoDB_collection_created' ,
+                       collection=self.collection, db=self.db_name)
 
-    def handle_logs(self, msgs):
+    def _parse_msg_for_MongoDB(self, msgs):
```
Comment (supriyopaul) [inline logagg/forwarders.py]: lower case fn name

File: logagg/forwarders.py [hunk 17]
```diff
@@ -25,95 +32,147 @@ class MongoDBForwarder(BaseForwarder):
...
         msgs_list = []
         #TODO: We need to do this by using iteration object.
         for msg in msgs:
-            msg_body = json.loads(msg.body.decode(encoding='utf-8',errors='strict'))
-            msg_body['id'] = msg_body.pop('id')
+            msg_body = json.loads(msg.body.decode(encoding='utf-8',
```
Comment (supriyopaul) [inline logagg/forwarders.py]: what happens on decode failure?

File: logagg/forwarders.py [hunk 18]
```diff
@@ -25,95 +32,147 @@ class MongoDBForwarder(BaseForwarder):
...
-            self.log.info("inserted %d msgs into mongodb" % (len(msgs)))
+            self.log.info('inserted_msgs_mongodb', num_msgs=len(msgs), type='metric')
+        except pymongo.errors.AutoReconnect(message='connection_to_mongoDB_failed'):
+            self._ensure_connection()
         except pymongo.errors.BulkWriteError as bwe:
-            self.log.exception('Write to mongo failed. Details: %s' % bwe.details)
-        except pymongo.errors.ServerSelectionTimeoutError:
-            self.log.exception('Push to databases and ack to nsq failed')
+            self.log.exception('bulk_write_to_mongoDB_failed', tb=bwe.details)
+            self._insert_1by1(msgs_list)
```
Comment (supriyopaul) [inline logagg/forwarders.py]: will double insertion fail?

File: logagg/forwarders.py [hunk 19]
```diff
@@ -25,95 +32,147 @@ class MongoDBForwarder(BaseForwarder):
...
+            msg_body['_id'] = msg_body.pop('id')
             msgs_list.append(msg_body)
+        return msgs_list
 
+    def _insert_1by1(self, records):
+        for r in records:
+            try:
+                self.collection.insert_one(r, ordered=False)
+            except pymongo.errors.OperationFailure as opfail:
+                self.log.exception('failed_to_insert_record_in_mongoDB',
```
Comment (supriyopaul) [inline logagg/forwarders.py]: explore using upsert in mongo

File: logagg/forwarders.py [hunk 20]
```diff
@@ -25,95 +32,147 @@ class MongoDBForwarder(BaseForwarder):
...
-            self.INFLUXDB_RECORDS = [record for record in self.INFLUXDB_RECORDS if record]
-            print 'record length %d' %(len(self.INFLUXDB_RECORDS))
-            try:
-                self.log.info('inserting the %d metrics into influxdb' % (len(self.INFLUXDB_RECORDS)))
-                self.influxdb_client.write_points(self.INFLUXDB_RECORDS)
-                self.log.info("inserted the metrics into influxdb %d" % (len(self.INFLUXDB_RECORDS)))
-                self.INFLUXDB_RECORDS = []
-            except (InfluxDBClientError, InfluxDBServerError) as e:
-                self.log.exception("failed to insert metric %s" % (self.INFLUXDB_RECORDS))
+        self.INFLUXDB_RECORDS = [record for record in self.INFLUXDB_RECORDS if record]
```
Comment (supriyopaul) [inline logagg/forwarders.py]: all caps var names are reserved for const vars. this is being mutated.

File: logagg/forwarders.py [hunk 21]
```diff
@@ -25,95 +32,147 @@ class MongoDBForwarder(BaseForwarder):
...
+        try:
+            self.log.debug('inserting_the_metrics_into_influxdb')
+            self.influxdb_client.write_points(self.INFLUXDB_RECORDS)
+            self.log.info('metrics_inserted_into_influxdb',
+                           length=len(self.INFLUXDB_RECORDS),
+                           type='metric')
+            self.INFLUXDB_RECORDS = []
+        except (InfluxDBClientError, InfluxDBServerError) as e:
+            self.log.exception('failed_to_insert metric',
+                                record=(self.INFLUXDB_RECORDS),
```
Comment (supriyopaul) [inline logagg/forwarders.py]: dont' need parantheses

File: logagg/util.py [hunk 22]
```diff
@@ -34,3 +35,59 @@ def start_daemon_thread(target, args=()):
...
     return th
 
+def serialize_dict_keys(d, prefix=""):
+    """ returns all the keys in a dictionary
+    >>> serialize_dict_keys({"a": {"b": {"c": 1, "b": 2} } })
+    ['a', 'a.b', 'a.b.c', 'a.b.b']
+    """
+
+    keys = []
+    for k,v in d.iteritems():
```
Comment (supriyopaul) [inline logagg/util.py]: please log issue for py3 compat

File: logagg/util.py [hunk 23]
```diff
@@ -34,3 +35,59 @@ def start_daemon_thread(target, args=()):
...
+    '''
+    fd = dict()
+    keys = serialize_dict_keys(d)
+    d = AttrDict(d)
+    for key in keys:
+        value = deepgetattr(d, key)
+        if isinstance(value, dict):
+            pass
+        else:
+            fd[key] = value
```
Comment (supriyopaul) [inline logagg/util.py]: test perf by benchmarking. can we make it more efficient. imp because it happens for every single msg

File: logagg/util.py [hunk 24]
```diff
@@ -34,3 +35,59 @@ def start_daemon_thread(target, args=()):
...
+        pass
+ 
+    try:
+        import unicodedata
+        unicodedata.numeric(s)
+        return True
+    except (TypeError, ValueError):
+        pass
+ 
+    return False
```
Comment (supriyopaul) [inline logagg/util.py]: This => def isnumber(x): return isinstance(ast.literal_eval(x), numbers.Number) ???

File: logagg/collector.py (hunk 2 above)
Comment (supriyopaul) [reply logagg/collector.py]: @prashanthellina  I was about to make _remove_redundancy_remove a seperate function but then I thought why do a separate iteration on the keys if we can complete it in one..?

File: README.md [hunk 25]
```diff
@@ -1,313 +1,358 @@
...
+    * `type` (str) - One of "log", "metric" (Is there one more type?)
+    * `level` (str) - Log level of the log line.
+    * `event` (str) - LOG event
+    * `error` (bool) - True if collection handler failed during processing
+    * `error_tb` (str) - Error traceback
 
-### Usage
 
-- After installation of the logagg module, we can perform operations in the python shell.
+---
```
Comment (prashanthellina) [inline README.md]: @supriyopaul  These hyphens looks odd. Don't believe they are required. Also, for the above fields, example values might help the user understand them better.

File: README.md [hunk 26]
```diff
@@ -1,313 +1,358 @@
...
-  --depth-limit-at-nsq DEPTH_LIMIT_AT_NSQ
-                        To limit the depth at nsq channel
-  --exception-logs-file EXCEPTION_LOGS_FILE
-                        If collector fails to publish messages to nsq, will
-                        write the logs to a file
-```
+    ```bash
+    $ logagg collect --help
+    ```
+    ##### or
```
Comment (prashanthellina) [inline README.md]: @supriyopaul  Hashes are a mechanism to specify header levels. Is this required here?

File: logagg/formatters.py [hunk 27]
```diff
@@ -9,12 +9,12 @@ def docker_log_file_driver(line):
     log = json.loads(json.loads(line)['msg'])
     if 'formatter' in log.get('extra'):
         return RawLog(dict(formatter=log.get('extra').get('formatter'),
-                            raw=log.get('message'),
+                            raw=str(log.get('message')),
```
Comment (supriyopaul) [inline logagg/formatters.py]: dangerous to force to str. decoding might fail.

File: logagg/forwarders.py [hunk 28]
```diff
@@ -25,95 +32,148 @@ class MongoDBForwarder(BaseForwarder):
...
+            msg_body['id'] = msg_body.pop('id')
+            msgs_list.append(msg_body)
 
-        self.log.info('Parsing of metrics started')
-        records = self.parse_msg_to_send_influxdb(msgs_list)
-        self.INFLUXDB_RECORDS.extend(records)
-        self.log.info('Parsing of metrics is completed')
+        self.log.debug('parsing_of_metrics_started')
+        records = self.parse_msg_for_influxDB(msgs_list)
+        self.influxDB_records.extend(records)
```
Comment (supriyopaul) [inline logagg/forwarders.py]: pep8 var name

File: logagg/util.py (hunk 22 above)
Comment (supriyopaul) [reply logagg/util.py]: done
